src/Build/ElementLocation/XmlElementWithLocation.cs

```diff
@@ -91,7 +91,8 @@ internal ElementLocation Location
             {
                 // Caching the element location object saves significant memory
                 XmlDocumentWithLocation ownerDocumentWithLocation = (XmlDocumentWithLocation)OwnerDocument;
-                if (!String.Equals(_elementLocation.File, ownerDocumentWithLocation.FullPath, StringComparison.OrdinalIgnoreCase))
+
+                if (ownerDocumentWithLocation.FullPathHasChanged && !String.Equals(_elementLocation.File, ownerDocumentWithLocation.FullPath, StringComparison.OrdinalIgnoreCase))
```

Comment:
Are these equals showing up? I'd expect this to be a reference equality check in the equal case?

Reply:
Without my change, `_elementLocation.File` is `String.Empty` for all elements and attributes while `OwnerDocument.FullPath` is `null` or a path.

With my change, we now check a bool first to see if we should compare the paths second.  And again, if the files were loaded from disk and haven't been renamed, the string comparison won't happen anymore and the `ElementLocation` objects won't be recreated

Reply:
I'm not sure I follow, under what circumstances would FullPath be different from ElementLocation.File and you wouldn't want to recreate the element location?

Reply:
If they are different, you definitely want to recreate ElementLocation.  However, without my change, we weren't setting FullPath prior to loading the document so all ElementLocations would have no `File` and then after loading, we'd set FullPath and after that any access to `Location` would recreate the `ElementLocation`.  If we set `FullPath` first and the file path doesn't change, the `ElementLocation` doesn't get recreated anymore.

The optimized case is evaluating a project from disk, it will no longer recreate ElementLocation the first time you access the property (which happens all over the place).

Reply:
I get the first change - what I don't get is why you need the FullPathHasChanged check? Is that indicated by the fact that the owner document and my "file" are different?

Reply:
I think the only purpose of FullPathHasChanged is to avoid the string equals in the location getters, in the case where full path does not change. If the full path changes, `FullPathHasChanged` gets stuck to true and all subsequent location getters get to execute the getters.

Reply:
Yeah we talked offline, it's to avoid the string comparison. In the traces I saw this isn't showing up *at all* in CPU time. This isn't doing a full string comparison, it's almost always just a reference equality check, and hence I think it can be removed.

Reply:
I have confirmed that `_elementLocation.File` and `ownerDocumentWithLocation.FullPath` are the same reference so when loading projects from disk this check adds trivial overhead.  I've reverted that commit.